[DevCenter] Update data plane to 2024-05-01-preview - breaking changes

am-lim opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

Breaking change PR, remove deprecation warnings. Update data plane cmdlets to 2024-05-01-preview. Adapted from previous Closed PR: https://github.com/Azure/azure-powershell/pull/25737

## Mandatory Checklist

- Please choose the target release of Azure PowerShell. (⚠️**Target release** is a different concept from **API readiness**. Please click below links for details.)
  - [x] [General release](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Public preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Private preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Engineering build](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] No need for a release

- [ ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For SDK-based development mode, update `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`.
        * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. 
    * For autorest-based development mode, include the changelog in the PR description.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** adjust version of module manually in pull request

Nickcandy commented:
Hi @am-lim , please fix the errors in static analysis ci: https://dev.azure.com/azclitools/public/_build/results?buildId=197064&view=logs&j=b9afb22f-0837-5d02-bc24-6c30aee08bcf&t=990f6f37-4a21-5927-092a-220f38a460e9

am-lim commented:
> Hi @am-lim , please fix the errors in static analysis ci: https://dev.azure.com/azclitools/public/_build/results?buildId=197064&view=logs&j=b9afb22f-0837-5d02-bc24-6c30aee08bcf&t=990f6f37-4a21-5927-092a-220f38a460e9

Hi @Nickcandy, fixed the example error. The rest of the errors are breaking change errors that we want to merge in for the breaking change release (Ignite release). We already added deprecation warnings in a previous release.

Nickcandy commented:
Our next release would be major release. So we would merge your pr in next sprint. For more details please see our milestone: https://github.com/Azure/azure-powershell/milestones

vidai-msft commented:
/azp run

am-lim commented:
> /azp run

@vidai-msft Hello, is there anything needed from our side? I verified that the analyze errors are the breaking changes we want to introduce.

vidai-msft commented on src/DevCenter/DevCenterData.AutoRest/custom/Get-AzDevCenterUserDevBoxCustomizationTaskLog.ps1:
```diff
@@ -0,0 +1,184 @@
+.Example
+{{ Add code here }}
+.Example
+{{ Add code here }}
```
Please fill in the examples.

vidai-msft commented on src/DevCenter/DevCenterData.AutoRest/custom/New-AzDevCenterUserDevBoxCustomizationGroup.ps1:
```diff
@@ -0,0 +1,194 @@
+.Example
+{{ Add code here }}
+.Example
+{{ Add code here }}
```
Please fill in the examples.

vidai-msft commented on src/DevCenter/DevCenterData.AutoRest/custom/Test-AzDevCenterUserDevBoxCustomizationTaskAction.ps1:
```diff
@@ -0,0 +1,180 @@
+.Example
+{{ Add code here }}
+.Example
+{{ Add code here }}
```
Please fill in the examples

vidai-msft commented on src/DevCenter/DevCenterData.AutoRest/docs/Deploy-AzDevCenterUserEnvironment.md:
```diff
@@ -195,7 +195,7 @@ Accept wildcard characters: False
 
 ### -ExpirationDate
-The time the expiration date will be triggered (UTC), after which the environment and associated resources will be deleted.
+The time the expiration date will be triggered (UTC), after which theenvironment and associated resources will be deleted.
```
```suggestion
The time the expiration date will be triggered (UTC), after which the environment and associated resources will be deleted.
```

vidai-msft commented:
@am-lim Please follow this doc https://github.com/Azure/azure-powershell/blob/main/documentation/Debugging-StaticAnalysis-Errors.md#breaking-changes to suppress the breaking change warning in the CI.

vidai-msft commented on src/DevCenter/DevCenterData.AutoRest/custom/Deploy-AzDevCenterUserEnvironment.ps1:
```diff
@@ -53,172 +55,177 @@ BODY <IEnvironment>: Properties of an environment.
 function Deploy-AzDevCenterUserEnvironment {
-  [OutputType([System.Boolean])]
-  [CmdletBinding(DefaultParameterSetName = 'CreateExpanded', PositionalBinding = $false, SupportsShouldProcess, ConfirmImpact = 'Medium')]
+  [OutputType([Microsoft.Azure.PowerShell.Cmdlets.DevCenterdata.Models.Api20240501Preview.IEnvironment])]
```
The output type will be updated from bool to IEnvironment. This is also considered as breaking change but I did not see it in the breaking change announcement. May I know if it would cause break to the customers?

vidai-msft commented on src/DevCenter/DevCenterData.AutoRest/custom/Get-AzDevCenterUserCatalog.ps1:
```diff
@@ -16,133 +16,146 @@
 function Get-AzDevCenterUserCatalog {
-  [OutputType([System.String], [Microsoft.Azure.PowerShell.Cmdlets.DevCenterdata.Models.Api20231001Preview.ICatalog])]
-  [CmdletBinding(DefaultParameterSetName = 'List', PositionalBinding = $false)]
+  [OutputType([Microsoft.Azure.PowerShell.Cmdlets.DevCenterdata.Models.Api20240501Preview.ICatalog])]
```
Here is another sample that it may cause breaking change to our customers and it does not have pre-announcement.

vidai-msft replied:
I see more ps1 files are with the upgraded api version as the OutputType than those in the pre-announced pr. Some of them don't have breaking change attributes. May I know if they won't cause break or if you have any justifications?

am-lim replied:
I missed these since they did not fail the Static Analysis. Are these still considered breaking changes? How do I check for these other than manually verifying each file?

vidai-msft replied:
Yes, basically these are breaking changes unless there is something wrong in the current output type. The standard to identify breaking change is if the change will break customers' custom scripts.
Take this as an example, the original output is either `String` or `Api20231001Preview.ICatalog` and it will be updated to `Api20240501Preview.ICatalog`. So when the new version is released, for some scenarios when they expect String as the output, the scripts will be broken as the output is an object. For the others, when they expect Api20231001Preview.ICatalog, now the real output is Api20240501Preview.ICatalog. If anything inside the Api20240501Preview.ICatalog changed, then the custom would also fail like a missing property or updated property name. Hope that helps.

am-lim replied:
@vidai-msft Is there a way to automatically check these breaking changes instead of manually verifying each file?

vidai-msft replied:
You can leverage the static analysis to help detect the breaking changes. But it may have some missing cases. If you are sure these are breaking changes, then please fill in the migration guide also even without preannouncement. Otherwise, we can ignore.

am-lim replied:
Thank you. I will close this PR and update these cmdlets to include a preannouncement since they are breaking changes. Where can I find the migration guide?

am-lim commented:
/azp run

vidai-msft commented:
/azp run

vidai-msft reviewed (approved):
LGTM
